feat: add `blas/base/wasm/zscal`

ShabiShett07 opened the pull request:
---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes. report:
  - task: lint_filenames status: passed
  - task: lint_editorconfig status: passed
  - task: lint_markdown status: passed
  - task: lint_package_json status: passed
  - task: lint_repl_help status: passed
  - task: lint_javascript_src status: passed
  - task: lint_javascript_cli status: na
  - task: lint_javascript_examples status: passed
  - task: lint_javascript_tests status: passed
  - task: lint_javascript_benchmarks status: passed
  - task: lint_python status: na
  - task: lint_r status: na
  - task: lint_c_src status: na
  - task: lint_c_examples status: na
  - task: lint_c_benchmarks status: na
  - task: lint_c_tests_fixtures status: na
  - task: lint_shell status: na
  - task: lint_typescript_declarations status: passed
  - task: lint_typescript_tests status: passed
  - task: lint_license_headers status: passed ---

Progresses #2039

## Description

> What is the purpose of this pull request?

This pull request:

-   Adds the package `blas/base/wasm/zscal`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   progresses #2039

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

ShabiShett07 committed:
feat: add blas/base/wasm/zscal

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

ShabiShett07 committed:
chore: update test cases

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

ShabiShett07 commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/zscal/README.md:
````diff
@@ -0,0 +1,380 @@
+var zscal = require( '@stdlib/blas/base/wasm/zscal' );
+```
+
+#### zscal.main( N, za, zx, strideX )
````
@ShabiShett07 Sorry to be pedantic, but since we are here, would you mind changing `za` to `alpha` and `zx` to `x` here and elsewhere in this package?

ShabiShett07 replied:
sure

ShabiShett07 committed:
chore: update variable names

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

ShabiShett07 commented:
@kgryte, I have changed the variable names

Commit:
bench: update variable handling

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Commit:
docs: update jsdoc

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Commit:
chore: use appropriate packages

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Commit:
docs: update jsdoc

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Commit:
chore: update repl.txt

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Commit:
docs: update jsdoc

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Commit:
chore: update markdown

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte committed:
docs: fix variable name

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/zscal/benchmark/benchmark.module.main.js:
```diff
@@ -0,0 +1,141 @@
+
+		// Reallocate the underlying memory to allow storing one vector and a complex number:
+		nb = bytesPerElement( 'complex128' );
+		mod.realloc( ((N*2)*nb)+nb );
+
+		// Define a pointer (i.e., byte offset) to the first vector element:
+		xptr = 0;
+
+		// Define a pointer to a complex number:
+		zptr = (N*2) * nb;
+
+		// Write random values to module memory:
+		mod.write( xptr, uniform( (N*2), -100.0, 100.0, options ) );
+		mod.write( zptr, new Float64Array( [ 1.0, 0.0 ] ) );
+
+		// Retrieve a DataView of module memory:
+		view = mod.view;
+
+		b.tic();
+		for ( i = 0; i < b.iterations; i++ ) {
+			mod.main( (N*2), zptr, xptr, 1 );
+			byteOffset = xptr + ( (i%(N*2))*nb );
```
You seem to be a bit confused regarding bytes offsets and the mixing of real and complex.

```suggestion
		mod.realloc( (N*nb)+nb );

		// Define a pointer (i.e., byte offset) to the first vector element:
		xptr = 0;

		// Define a pointer to a complex number:
		zptr = N * nb;

		// Write random values to module memory:
		mod.write( xptr, uniform( (N*2), -100.0, 100.0, options ) );
		mod.write( zptr, new Float64Array( [ 1.0, 0.0 ] ) );

		// Retrieve a DataView of module memory:
		view = mod.view;

		b.tic();
		for ( i = 0; i < b.iterations; i++ ) {
			mod.main( N, zptr, xptr, 1 );
			byteOffset = xptr + ( (i%N)*nb );
```

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/zscal/benchmark/benchmark.module.ndarray.js:
```diff
@@ -0,0 +1,141 @@
+
+		// Reallocate the underlying memory to allow storing one vector and a complex number:
+		nb = bytesPerElement( 'complex128' );
+		mod.realloc( ((N*2)*nb)+nb );
```
Same changes here. You are over-allocating by 2x. `nb` is the number of bytes per element for `complex128`, not `float64`.

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/zscal/docs/repl.txt:
```diff
@@ -0,0 +1,549 @@
+    > {{alias}}.main( 2, alpha, x, 1 )
+    <Complex128Array>[ -3.0, 4.0, -5.0, 10.0 ]
+
+    // Advanced indexing:
```
```suggestion
    // Using `N` and stride parameters:
```

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/zscal/docs/repl.txt:
```diff
@@ -0,0 +1,549 @@
+    // Using typed array views:
+    > var x0 = new {{alias:@stdlib/array/complex128}}( [ 1.0, 2.0, 3.0, 4.0, 5.0, 6.0 ] );
+    > var x1 = new {{alias:@stdlib/array/complex128}}( x0.buffer, x0.BYTES_PER_ELEMENT*1 );
+    > var alpha = new {{alias:@stdlib/complex/float64/ctor}}( 2.0, 2.0 );
```
```suggestion
    > alpha = new {{alias:@stdlib/complex/float64/ctor}}( 2.0, 2.0 );
```

`alpha` has already been declared above.

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/zscal/docs/repl.txt:
```diff
@@ -0,0 +1,549 @@
+        Number of indexed elements.
+
+    alpha: Complex128
+        Complex constant.
```
```suggestion
        Scalar constant.
```

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/zscal/docs/repl.txt:
```diff
@@ -0,0 +1,549 @@
+        Number of indexed elements.
+
+    alpha: Complex128
+        Complex constant.
```
```suggestion
        Scalar constant.
```

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/zscal/docs/repl.txt:
```diff
@@ -0,0 +1,549 @@
+        Number of indexed elements.
+
+    ap: integer
+        Pointer (i.e., byte offset) to a scalar complex constant.
```
```suggestion
        Pointer (i.e., byte offset) to a scalar constant.
```

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/zscal/docs/repl.txt:
```diff
@@ -0,0 +1,549 @@
+        Number of indexed elements.
+
+    ap: integer
+        Pointer (i.e., byte offset) to a scalar complex constant.
```
```suggestion
        Pointer (i.e., byte offset) to a scalar constant.
```

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/zscal/test/test.module.main.js:
```diff
@@ -0,0 +1,295 @@
+	mod.write( xp, new Complex128Array( [ 0.3, 0.1, 0.5, 0.0, 0.0, 0.5 ] ) );
+	mod.write( ap, new Float64Array( [ 0.4, -0.7 ] ) );
+
+	out = mod.main( 0, ap, xp, 1 );
```
This test is incorrect. You need to actually confirm that the original array is unchanged. This test does nothing of the sort. For an idea of what is expected, see `blas/base/wasm/daxpy/test/test.module.main.js`.

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/zscal/test/test.module.main.js:
```diff
@@ -0,0 +1,295 @@
+	t.end();
+});
+
+tape( 'a module instance has a `main` method which returns a reference to the input array', function test( t ) {
```
It doesn't return a `reference`. It returns a `pointer`. See the Wasm implementation for `daxpy`.

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/zscal/test/test.module.ndarray.js:
```diff
@@ -0,0 +1,423 @@
+	t.end();
+});
+
+tape( 'a module instance has an `ndarray` method which returns a reference to the input array', function test( t ) {
```
It doesn't return a `reference`. It returns a `pointer`. See the Wasm implementation for `daxpy`.

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/zscal/test/test.module.ndarray.js:
```diff
@@ -0,0 +1,423 @@
+	mod.write( ap, new Float64Array( [ 0.4, -0.7 ] ) );
+
+	out = mod.ndarray( 0, ap, xp, 1, 0 );
+	t.strictEqual( out, xp, 'returns expected value' );
```
Same comment. You are not testing that the array is left unchanged.

kgryte reviewed (changes requested):
Left initial comments. This PR is mostly in good shape. The most glaring defect is in the tests.

Commit:
chore: update memory allocation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Commit:
chore: update repl.txt

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Commit:
fix: update test cases

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

ShabiShett07 committed:
chore: update repl.txt

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Shabareesh Shetty <139731143+ShabiShett07@users.noreply.github.com>

ShabiShett07 commented:
@kgryte, ready for review

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/zscal/test/test.module.ndarray.js:
```diff
@@ -0,0 +1,431 @@
+	mod.initializeSync();
+
+	xp = 0;
+	ap = 24;
```
```suggestion
	ap = 48;
```

kgryte committed:
test: fix pointer

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (approved):
LGTM
